python/convert.c

```diff
@@ -0,0 +1,327 @@
+      return val1.double_val == val2.double_val;
+    case UPB_TYPE_STRING:
+    case UPB_TYPE_BYTES:
+      return val1.str_val.size == val2.str_val.size &&
+          memcmp(val1.str_val.data, val2.str_val.data, val1.str_val.size) == 0;
+    case UPB_TYPE_MESSAGE:
+      return PyUpb_Message_IsEqual(val1.msg_val, val2.msg_val,
+                                   upb_fielddef_msgsubdef(f));
+    default:
+      return false;
+  }
+}
+
+bool PyUpb_Map_IsEqual(const upb_map *map1, const upb_map *map2,
+                       const upb_fielddef *f) {
+  assert(upb_fielddef_ismap(f));
+  const upb_msgdef *entry_m = upb_fielddef_msgsubdef(f);
+  const upb_fielddef *val_f = upb_msgdef_field(entry_m, 1);
+  size_t iter = UPB_MAP_BEGIN;
+  size_t size1 = map1 ? upb_map_size(map1) : 0;
+  size_t size2 = map2 ? upb_map_size(map2) : 0;
+
+  if (map1 == map2) return true;
+  if (size1 != size2) return false;
+  if (size1 == 0) return true;
+
+  while (upb_mapiter_next(map1, &iter)) {
+    upb_msgval key = upb_mapiter_key(map1, iter);
+    upb_msgval val1 = upb_mapiter_value(map1, iter);
+    upb_msgval val2;
+    if (!upb_map_get(map2, key, &val2)) return false;
+    if (!PyUpb_ValueEq(val1, val2, val_f)) return false;
+  }
+
+  return true;
+}
+
+static bool PyUpb_ArrayElem_IsEqual(const upb_array *arr1,
+                                    const upb_array *arr2, size_t i,
+                                    const upb_fielddef *f) {
+  upb_msgval val1 = upb_array_get(arr1, i);
+  upb_msgval val2 = upb_array_get(arr2, i);
+  return PyUpb_ValueEq(val1, val2, f);
+}
+
+bool PyUpb_Array_IsEqual(const upb_array *arr1, const upb_array *arr2,
+                         const upb_fielddef *f) {
+  assert(upb_fielddef_isseq(f) && !upb_fielddef_ismap(f));
+  size_t n1 = arr1 ? upb_array_size(arr1) : 0;
+  size_t n2 = arr2 ? upb_array_size(arr2) : 0;
+
+  if (arr1 == arr2) return true;
+  if (n1 != n2) return false;
+
+  // Half the length rounded up.  Important: the empty list rounds to 0.
+  size_t half = (n1 + 1) / 2;
+
+  // Search from the ends-in.  We expect differences to more quickly manifest
+  // at the ends than in the middle.  We may compare the most middle element
+  // twice.
+  for (size_t i = 0; i < half; i++) {
+    if (!PyUpb_ArrayElem_IsEqual(arr1, arr2, i, f)) return false;
+    if (!PyUpb_ArrayElem_IsEqual(arr1, arr2, n1 - 1 - i, f)) return false;
+  }
+
+  return true;
+}
+
+bool PyUpb_Message_IsEqual(const upb_msg *msg1, const upb_msg *msg2,
+                           const upb_msgdef *m) {
+  size_t iter1 = UPB_MSG_BEGIN;
+  size_t iter2 = UPB_MSG_BEGIN;
+
+  if (msg1 == msg2) return true;
+
+  while (true) {
+    const upb_fielddef *f1, *f2;
+    upb_msgval val1, val2;
+    bool ok1 = msg1 && upb_msg_next(msg1, m, NULL, &f1, &val1, &iter1);
+    bool ok2 = msg2 && upb_msg_next(msg2, m, NULL, &f2, &val2, &iter2);
+    if (ok1 != ok2) return false;
+    if (!ok1) break;  // Both messages are at end.
+    if (f1 != f2) return false;
```

Comment:
can you add a comment here saying "if the fields have different types" I spent a while puzzling over why the rest was useful is there was already pointer identity

Reply:
Added a comment (my explanation for this is somewhat different than your suggestion).